core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -293,8 +293,15 @@ public void commit() {
             taskOps.commit(base, updated.withUUID());
           });
 
+    } catch (CommitFailedException commitFailedException) {
+      // We have an acknowledged failure from the Catalog. We are confident that the commit has not been applied
+      Exceptions.suppressAndThrow(commitFailedException, this::cleanAll);
     } catch (RuntimeException e) {
-      Exceptions.suppressAndThrow(e, this::cleanAll);
+      LOG.error("Cannot determine whether the commit was successful or not, the underlying data files may or " +
```

Comment:
I am afraid this will be dangerous. This means we have to update all places that throw a specific exception like `AlreadyExistsException` to now throw `CommitFailedException`. This is error-prone and we lose details that will be helpful. We could throw `CommitFailedException` and add the specific exception as a cause but the first point still holds.

Also, an error log message will most likely be ignored by the user. This is a case where we really want to propagate as much info as possible.

Would it make sense to introduce a new exception type? E.g., `UnknownCommitStateException`? That way, we can keep the existing logic for handling exceptions except cases where we really don't know what happens.

Reply:
Thoughts, @RussellSpitzer @pvary @rymurr @danielcweeks @Parth-Brahmbhatt?

Reply:
cc @omalley @rdblue @shardulm94 @rdsr too

Reply:
I agree wtih @aokolnychyi here, I think changing the contract for the table ops is dangerous, especially as the change reduces the information available to the user.

I think Anton is right, rather than saying "`CommitFailedException` is the only valid known failure mode" we should probably say "`NewExceptionTypeX` means the commit failed in an unknown way and the user may be required to do some immediate intervention"

Reply:
I think that's fine, I just want to make sure every catalog implements it then, if we always treat all non CME excepts as unknown commit state we know we are never deleting state we need. If we instead do it in the opposite direction we need to be extra sure that all exceptions that may have a unknown state are marked that way.

Reply:
Switched around the implementation so now it uses a specific exception CommitStateUnknownException. We need to be care going forward with reviews of catalog code to make sure developers properly throw this.

Reply:
Thanks @aokolnychyi . I'll have a look at this by tomorrow!

Reply:
I should also mention that we handle this internally by checking in our `TableOperations` when there is a transport exception. We just try to load the table from the metastore and see if it has the new table metadata location. If so, then we return success. That's a better initial fix than updating the logic here.

Reply:
The first scenario you propose is actually the most dangerous without this change, another client could beat the deletion code and end up doing another commit on top of the files as they are being removed. This then leads to an unrecoverable situation since additional writes and operations will be based of of files that are removed.

Reply:
You're right that there is a race condition there. But it is far more unlikely that there is an unknown state problem followed by an immediate race commit. It's more likely that deleting the metadata file immediately will cause other commits to fail.

I think that it's reasonable to do this if the committer is idempotent, but otherwise I would not recommend doing it by default. We'd be trading the right behavior almost all of the time for the right behavior in an extreme case.

Reply:
I actually am not sure that is the case here, we are in the situation were one client, clientA has somehow lost the ability to confirm, but i'm not sure it's more likely that client B would also lose the ability to confirm. I guess it depends on if you believe a server-level failure is more likely than a client-level failure.

In my experience it is much more likely that single client has connection issues than all clients have connection issues, but ymmv

Reply:
Also you'll note we do have the check here in table operations as well, we are checking the metastore (or at least attempting to) immediately after failure to try to avoid going down this path.

Reply:
I agree it is a valid concern when clients don't handle `UnknownStateExceptions` and blindly retry. However, we can try to address it in another way. For example, if users have idempotent requirements, they either should make their retry mechanisms smarter and be aware of `UnknownCommitStateException` or use something that guarantees the same batch won't be committed twice. We already handle such cases in Structured Streaming. I hope we do the same in Flink jobs. We can generalize this approach and make Iceberg handle such cases in batch writes too.

I feel the current approach is extremely dangerous. If we delete a file that was actually committed, we corrupt the table and require a manual intervention. Unfortunately, I've seen this quite frequently. Also, getting a reply from the metastore or checking the commit status may take a substantial amount of time allowing a concurrent operation to succeed. That's the worst what can happen as we silently corrupt the table and we will detect it only while querying the table. I had to fix such a case and we were simply lucky we detected this only after a couple of days while rewriting manifests.

Reply:
For example, we could expose a hook to provide logic for checking whether a particular operation is already committed.  Kind of what we have in Structured Streaming where Spark gives us queryId and batchId.

Alternatively, blocking indefinitely may be an option. However, we keep only last 100 metadata files (by default) so the longer we wait, the less likely we will correctly check if a commit was applied.

Reply:
I thought more about this today and I am still convinced deleting files in an attempt to fail other concurrent operations (which is not guaranteed) and requiring a manual intervention to restore the table is NOT something we should do due to the described consequences and cases that we won't handle.

I also like throwing an exception rather than hanging forever as this seems to be more descriptive. I think the error message Russell throws right now is pretty elaborate and it should be propagated as a cause of `SparkException`. As a user, I'd prefer my job to fail with `UnknownCommitStateException` rather than with a job timeout exception and no signs of what went wrong.

Internally, we offer job retry mechanisms but we explicitly warn our users that they may enable it only if their jobs are idempotent. We can try to reduce the chances of retries but I don't think we can solve that completely when we loose the metastore connectivity. I think it is reasonable to explore how we can skip already committed batch operations like we do in Structured Streaming but that can be done in a follow-up PR.

It may be also a good call to check the commit state more than once for some reasonable time frame but it shouldn't be for too long to avoid job timeouts.

Reply:
What do you think, @RussellSpitzer @pvary @rdblue @shardulm94 @marton-bod @Parth-Brahmbhatt?

Reply:
This is inline with what I was thinking

I am strongly against deleting files when we are not sure of the state.

I prefer the explicit exception because I also think it's probably the best way to actually get an end user information about the failure and allow them or a framework to automatically respond to it. I worry that doing a infinite blocking or even a very extended retry can lead to another user mechanism terminating the job, in that case they would have to directly inspect the logs to know that something may have failed.

Reply:
Any reason we can't make this configurable so the client is forced to indicate if they are idempotent or not explicitly?

For non idempotent clients we should keep the current behavior with limited and configurable retries to check if a snapshot was committed before issuing final snapshot deletion. If the client indicates they are idempotent we can avoid deleting and throw the exception as proposed in this PR.

Reply:
I'll be okay if we want to make this configurable. How will the config look like? Will it be a table property?

The question is what should be the default behavior. I'd say no matter whether clients are idempotent or not, it is safer not to delete files due to the potential table corruption. Adding the same data twice is not as critical, in my view.

Reply:
The current behavior is very dangerous, it can cause changes to the table which are unrecoverable. I don't think it is a viable option and I would strongly recommend we don't make it the default even if we retry before doing it. From our own internal investigations and other reports I've heard, this behavior has actually been cropping up in a number of places but it's rather hard to track down.

Reply:
I don't think the property should be persisted at table level as a single table could be written from both idempotent and non idempotent clients (flink writer (idempotent) + background a compaction process (nonidempotent)). We can either expose the API at Table or PendingUpdate level. 

In absence of more external use cases it is hard to say what is worst , corrupting a table so nothing can read or write until manual intervention or having duplicate data that can go undetected. I would pick duplicate data based on general use cases that I have seen so if we are voting, I would vote to default the behavior to assuming idempotent client.

Reply:
An API on `PendingUpdate` sounds good to me.

One more reason why I'd vote for not deleting by default is due to this situation:
> Also, getting a reply from the metastore or checking the commit status may take a substantial amount of time allowing a concurrent operation to succeed (if the lock expires). That's the worst what can happen as we will silently corrupt the table and will detect it only while querying a specific portion of the table. I had to fix such a case and we were simply lucky we found this out only after a couple of days while rewriting manifests.

Reply:
@Parth-Brahmbhatt one thing to remember is that in our current code we actually get both data duplication and corrupt tables, it just depends when the failure occurs. Ie if you fail to check the metastore and lose connection to the file system, you won't clean up and will still throw an error. Or if say the commit is successful but for some other reason the job fails (OOM, Interrupt, power loss, I spill my coke on the server) we still end up with a failed job which if we retry will duplicate data / work.


So you can still get data duplication if a container loses connectivity to the network after committing.

Reply:
I think there are a couple of issues here that we should address separately.

First, I agree that it is a good idea to add a way to signal that a commit is idempotent. Some operations are already idempotent, like file rewrites because we validate that all of the rewritten files are still part of the table. If we update paths that guarantee idempotent commits to signal this and handle `UnknownCommitStateException`, then we really reduce the incidence of the problem.

Second, I think we still need to agree on the default behavior. While I really don't like the idea of allowing a retry that will write duplicate data, I think that this has convinced me that silently duplicating data is a better outcome.

In our environment, we use versioned buckets so we can always un-delete metadata files that are missing, but that's not always the case. If those files are actually gone, then it is much worse because you have missing data and don't know which data files were in the missing commit without a lot of work. I think this problem is worse than the duplicate data.

A second compelling argument for changing the default is that deleting the metadata files leaks the problem to other writers. All concurrent processes are blocked if a table is broken, rather than blocking just a single writer.

In the end, I think that the right thing is to not delete the files and to throw `UnknownCommitStateException` as proposed. That handles interactive cases and also makes it so schedulers can handle a job failure by blocking just a single workflow and not all workflows operating on a table. And idempotent jobs should not be affected.

Reply:
I would go for API level flag that the change is idempotent since this is more dependent on the client / action then the actual table.

Also I would go for a few connection retries and then throwing the `UnknownCommitStateException` if we are not able to determine the status of the commit. We should fail fast as soon as possible, so the user is able to mitigate the issue. If they ignore the exception it is better to have a state where we can recover so I would keep the files in case we are not able to handle the error cleanly.